Build failure: zsync

## Steps To Reproduce

Steps to reproduce the behavior:

1. build *zsync*

## Build log

<!-- insert build log in code block in collapsable section -->

<details>

<summary>Build Log</summary>

```
error: builder for '/nix/store/rsvzh1sz4ln7j4jvj0p2g04kh8xmv384-zsync-0.6.2.drv' failed with exit code 2;
       last 25 log lines:
       > zsync.c:928:42: warning: comparison of integer expressions of different signedness: 'off_t' {aka 'long int'} and 'uLong' {aka 'long unsigned int'} [-Wsign-compare]
       >   928 |     if (zr->strm.total_in == 0 || offset != zr->strm.total_in) {
       >       |                                          ^~
       > zsync.c:930:43: error: passing argument 4 of 'zsync_configure_zstream_for_zdata' from incompatible pointer type [-Wincompatible-pointer-types]
       >   930 |                                           &(zr->outoffset));
       >       |                                           ^~~~~~~~~~~~~~~~
       >       |                                           |
       >       |                                           off_t * {aka long int *}
       > zsync.c:771:65: note: expected 'long long int *' but argument is of type 'off_t *' {aka 'long int *'}
       >   771 |                                        long zoffset, long long *poutoffset) {
       >       |                                                      ~~~~~~~~~~~^~~~~~~~~~
       > zsync.c:942:26: warning: assignment discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]
       >   942 |         zr->strm.next_in = buf;
       >       |                          ^
       > zsync.c:954:17: warning: this statement may fall through [-Wimplicit-fallthrough=]
       >   954 |             eoz = 1;
       >       |             ~~~~^~~
       > zsync.c:955:9: note: here
       >   955 |         case Z_BUF_ERROR:
       >       |         ^~~~
       > make[2]: *** [Makefile:244: zsync.o] Error 1
       > make[2]: Leaving directory '/build/zsync-0.6.2/libzsync'
       > make[1]: *** [Makefile:434: all-recursive] Error 1
       > make[1]: Leaving directory '/build/zsync-0.6.2'
       > make: *** [Makefile:279: all] Error 2
       For full logs, run 'nix log /nix/store/rsvzh1sz4ln7j4jvj0p2g04kh8xmv384-zsync-0.6.2.drv'.
```

</details>

## Additional context

<!-- Add any other context about the problem here. -->
Found while compiling quickemu.

## Metadata

<!-- Please insert the output of running `nix-shell -p nix-info --run "nix-info -m"` below this line -->
- system: `"x86_64-linux"`
 - host os: `Linux 6.6.66, NixOS, 25.05 (Warbler), 25.05.20241219.d70bd19`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Lix, like Nix) 2.91.1`

## Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->
@viric 
@FedX-sudo @flexiondotorg (From quickemu maintainers)

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/zs/zsync/package.nix
+++ b/pkgs/by-name/zs/zsync/package.nix
@@ -1,25 +1,45 @@
 {
   lib,
   stdenv,
-  fetchurl,
+  fetchFromGitHub,
+  autoreconfHook,
 }:
 
 stdenv.mkDerivation rec {
   pname = "zsync";
-  version = "0.6.2";
+  version = "0.6.2-unstable-2017-04-25";
 
-  src = fetchurl {
-    url = "http://zsync.moria.org.uk/download/zsync-${version}.tar.bz2";
-    hash = "sha256-C51TQzOHqk8EY0psY6XvqCAwcPIpivcqcF+b492mWvI=";
+  src = fetchFromGitHub {
+    owner = "cph6";
+    repo = "zsync";
+    rev = "6cfe374f8f2310cbd624664ca98e5bb28244ba7a";
+    hash = "sha256-SnCzNDMyhMx+2JmgsrjtYDa31Ki1EWix9iBfaduDnro=";
   };
 
-  env = lib.optionalAttrs stdenv.cc.isClang {
-    # Suppress error "call to undeclared library function 'strcasecmp'" during compilation.
-    # The function is found by the linker correctly, so this doesn't introduce any issues.
-    NIX_CFLAGS_COMPILE = " -Wno-implicit-function-declaration";
-  };
+  sourceRoot = "${src.name}/c";
+
+  patches = [
+    ./remove-inexisting-rsumtest.patch
+    ./read-blocksums-declaration-fix.patch
+  ];
 
   makeFlags = [ "AR=${stdenv.cc.bintools.targetPrefix}ar" ];
+
+  # Suppress error "call to undeclared library function 'strcasecmp'" during compilation.
+  # The function is found by the linker correctly, so this doesn't introduce any issues.
+  # Also supress errors that come from incompatible pointer types due to GCC 14 changes.
+  env.NIX_CFLAGS_COMPILE = toString (
+    lib.optionals stdenv.cc.isClang [
+      "-Wno-implicit-function-declaration"
+    ]
+    ++ lib.optionals stdenv.cc.isGNU [
+      "-Wno-error=incompatible-pointer-types"
+    ]
+  );
+
+  nativeBuildInputs = [
+    autoreconfHook
+  ];
 
   meta = with lib; {
     homepage = "http://zsync.moria.org.uk/";

--- a/pkgs/by-name/zs/zsync/package.nix
+++ b/pkgs/by-name/zs/zsync/package.nix
@@ -41,11 +41,14 @@ stdenv.mkDerivation rec {
     autoreconfHook
   ];
 
-  meta = with lib; {
-    homepage = "http://zsync.moria.org.uk/";
+  meta = {
+    homepage = "https://github.com/cph6/zsync";
     description = "File distribution system using the rsync algorithm";
-    license = licenses.artistic2;
-    maintainers = with maintainers; [ viric ];
-    platforms = with platforms; all;
+    license = lib.licenses.artistic2;
+    maintainers = with lib.maintainers; [
+      viric
+      ryand56
+    ];
+    platforms = with lib.platforms; all;
   };
 }

--- a/pkgs/by-name/zs/zsync/read-blocksums-declaration-fix.patch
+++ b/pkgs/by-name/zs/zsync/read-blocksums-declaration-fix.patch
@@ -1,0 +1,13 @@
+diff --git a/libzsync/zsync.c b/libzsync/zsync.c
+index 793a426..783c349 100644
+--- a/libzsync/zsync.c
++++ b/libzsync/zsync.c
+@@ -116,7 +116,7 @@ struct zsync_state {
+ };
+ 
+ static int zsync_read_blocksums(struct zsync_state *zs, FILE * f,
+-                                int rsum_bytes, int checksum_bytes,
++                                int rsum_bytes, unsigned int checksum_bytes,
+                                 int seq_matches);
+ static int zsync_sha1(struct zsync_state *zs, int fh);
+ static int zsync_recompress(struct zsync_state *zs);

--- a/pkgs/by-name/zs/zsync/remove-inexisting-rsumtest.patch
+++ b/pkgs/by-name/zs/zsync/remove-inexisting-rsumtest.patch
@@ -1,0 +1,17 @@
+diff --git a/c/librcksum/Makefile.am b/c/librcksum/Makefile.am
+index 0216e49..f4fce72 100644
+--- a/librcksum/Makefile.am
++++ b/librcksum/Makefile.am
+@@ -2,9 +2,8 @@
+ 
+ noinst_LIBRARIES = librcksum.a
+ 
+-TESTS = md4test rsumtest
+-noinst_PROGRAMS = md4test rsumtest
++TESTS = md4test
++noinst_PROGRAMS = md4test
+ 
+ md4test_SOURCES = md4test.c md4.h md4.c
+-rsumtest_SOURCES = rsum.c rsumtest.c hash.c range.c state.c md4.c ../progress.c
+ librcksum_a_SOURCES = internal.h rcksum.h md4.h rsum.c hash.c state.c range.c md4.c
+

